improve bootstrap change-tracking system

onur-ozkan opened the pull request:
This PR aims to improve how change-tracking system works for bootstrap changes by doing the followings:

- Enforce embedding directive informations about the changes on `bootstrap/src/lib.rs`.
- Give more informative change inputs on the terminal (https://github.com/rust-lang/rust/pull/117815#discussion_r1390239657).
- Avoid spamming the change informations(by reading and creating `.last-warned-change-id` under build output dir).

see the zulip conversation for more details: https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/config.2Etoml.20change.20tracking

cc @RalfJung

onur-ozkan committed:
print the change warnings once for per id

Signed-off-by: onur-ozkan <work@onurozkan.dev>

RalfJung commented on triagebot.toml:
```diff
@@ -583,11 +583,11 @@ message = "Third-party dependency whitelist may have been modified! You must ens
 
 [mentions."src/bootstrap/src/core/config"]
-message = "This PR modifies `src/bootstrap/src/core/config`. If appropriate, please also update `CONFIG_CHANGE_HISTORY` in `src/bootstrap/src/lib.rs` and `change-id` in `config.example.toml`."
+message = "This PR modifies `src/bootstrap/src/core/config`. If appropriate, please provide an explanation of the modification in the PR description, update `CONFIG_CHANGE_HISTORY` in `src/bootstrap/src/lib.rs` and `change-id` in `config.example.toml`."
```
It's not about an explanation of the modification. It's about explaining *what rustc devs should change in their config.toml*. Keep in mind who these messages are addressing.

RalfJung replied:
Also this has some overlap with https://github.com/rust-lang/rust/pull/117809

RalfJung commented on src/bootstrap/src/bin/main.rs:
```diff
@@ -132,6 +140,8 @@ fn check_version(config: &Config) -> Option<String> {
                 "update `config.toml` to use `change-id = {latest_config_id}` instead"
             ));
+
+            t!(fs::write(warned_id_path, id.to_string()));
```
Can you detect if this is running in a terminal or not? When it is not in a terminal we probably don't want to update this file.

onur-ozkan replied:
> "running in a terminal or not"
You mean if it's CI environment or not?

RalfJung replied:
No, I mean running in a terminal (as in the `is_terminal` function). x.py is also often executed by rust-analyzer, to give one example.

onur-ozkan replied:
Got it, thanks.

https://github.com/rust-lang/rust/pull/117815/commits/89829bb83215799e8c99bdda0b753b59e73aa119 should do the work.

albertlarsan68 commented:
I have a few questions:

1. Does `x clean` clear the file? I would prefer if it did.
2. Is the file created/updated only upon printing the second message? It would be bad if it was only shown once at the start, eclipsed by a few walls of texts, then never show again until the next time the field is updated.
3. Maybe add a config (off by default) to always show it?

Nilstrieb commented:
I think a problem with this system is that there are several audiences. First, there are rust-lang/rust developers. Then there are distros packaging Rust. These two audiences have very different needs. I would think that distros want to know about the all the breakage, while rust-lang/rust devs don't really care.
Maybe the message should look at the profile and only display the messages for `profile = "dist"`? Or maybe have two lists, those with less critical changes for distros and those with super important critical changes for everyone. Although I'm not sure whether super important critical changes for everyone even exist, or should ever exist.

onur-ozkan commented:
```rs
// PR ID(to provide PR link for more input), log prefix, change summary
pub const CONFIG_CHANGE_HISTORY: &[(usize, &str, &str)] = &[
    (115898, "NOTE", "Implementation of this change-tracking system. Ignore this."),
    (116998, "INFO", "Removed android-ndk r15 support in favor of android-ndk r25b."),
    (117435, "INFO", "New option `rust.parallel-compiler` added to config.toml."),
    (116881, "WARNING", "Default value of `download-ci-llvm` was changed for `codegen` profile.")
];
```

Any thoughts on this(ignore the ugliness of the code)? This approach can automatically force PR authors to add change summaries.

albertlarsan68 commented:
Would be good if the severity was an Enum, so that different actions could be taken based on if it should be fatal or not (really breaking or just an information, show just once (at the end of the build preferably), or always)

RalfJung commented:
I would expect distros to check the changelog. So the changelog should have a section on `config.toml` changes. I don't think this system needs to be concerned with distros at all.

> Any thoughts on this(ignore the ugliness of the code)? This approach can automatically force PR authors to add change summaries.

Yeah, that seems nice!

onur-ozkan committed:
refactor change-tracking implementation

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented on src/bootstrap/src/bin/main.rs:
```diff
@@ -108,35 +108,44 @@ fn check_version(config: &Config) -> Option<String> {
+            msg.push_str("There have been changes to x.py since you last updated:\n");
+
+            for change in changes {
+                msg.push_str(&format!("  [{}] {}\n", change.severity.to_string(), change.summary));
+                msg.push_str(&format!(
+                    "    - PR Link https://github.com/rust-lang/rust/pull/{}\n",
+                    change.change_id
+                ));
```
This change results this format:

```sh
Building bootstrap
   Compiling bootstrap v0.0.0 (/home/nimda/devspace/onur-ozkan/rust/src/bootstrap)
    Finished dev [unoptimized] target(s) in 4.85s
There have been changes to x.py since you last updated:
  [INFO] New option `rust.parallel-compiler` added to config.toml.
    - PR Link https://github.com/rust-lang/rust/pull/117435
  [INFO] Default value of `download-ci-llvm` was changed for `codegen` profile.
    - PR Link https://github.com/rust-lang/rust/pull/116881
note: to silence this warning, update `config.toml` to use `change-id = 116881` instead
Building stage0 tool tidy (x86_64-unknown-linux-gnu)

```

onur-ozkan committed:
remove .last-warned-change-id on `x clean`

Signed-off-by: onur-ozkan <work@onurozkan.dev>

Kobzol commented on src/bootstrap/src/lib.rs:
```diff
@@ -69,16 +69,59 @@ const LLVM_TOOLS: &[&str] = &[
+
+#[derive(Clone, Debug)]
+pub enum Severity {
+    Info,
```
It would be useful to document here what do these levels mean, and perhaps provide some examples (like "info = giving notice about new options, no need to modify your local config", "warning = you might need to modify your local config").

onur-ozkan replied:
Does https://github.com/rust-lang/rust/pull/117815/commits/a4f05c6e0e54f879de17b0c2b8deecac17f87fe5 help?

onur-ozkan replied:
> and perhaps provide some examples

The existing change list is already a good example I think.

Kobzol replied:
Yes, great! Thank you.

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6835176097/job/18588673686) [(plain)](https://github.com/rust-lang/rust/commit/426a2f8821040851198e938bfe0940a759eb2842/checks/18588673686/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=onur-ozkan
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_5247b69c-23a9-4ffe-b38a-dd4692509af2
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=f184488d744a5f929eb817e15e38f9dac49a4bf6
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_5247b69c-23a9-4ffe-b38a-dd4692509af2
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_5247b69c-23a9-4ffe-b38a-dd4692509af2
GITHUB_TRIGGERING_ACTOR=onur-ozkan
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117815/merge
GITHUB_WORKFLOW_SHA=f184488d744a5f929eb817e15e38f9dac49a4bf6
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
```

</details>

onur-ozkan committed:
write .last-warned-change-id only if environment is tty

As the .last-warned-change-id is only used for change tracking,
we don't need to generate/write it outside of the tty.
Otherwise, rust-analyzer could create this file, and developers
wouldn't be able to see the bootstrap change alerts, assuming
that they have already seen them.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented:
> Does x clean clear the file? I would prefer if it did.

Yes it does clear the file.

> It would be bad if it was only shown once at the start, eclipsed by a few walls of texts, then never show again until the next time the field is updated.

We print these logs twice (2nd print is at the end of bootstrap) to ensure devs will see them.

> Maybe add a config (off by default) to always show it?

As this option doesn't exists in bootstrap for old branches, builds will fail on recent-to-old checkouts.

Printing these logs once for per initialization (after `x clean`, or on the fresh repository, or after `rm -rf build`) is fine in my opinion.

RalfJung commented on src/bootstrap/src/lib.rs:
```diff
@@ -69,16 +69,61 @@ const LLVM_TOOLS: &[&str] = &[
+    /// Used when upstream bootstrap configurations continue working as before.
+    Info,
+    /// Used when upstream bootstrap configurations change, and it may
+    /// require developers to update their build configuration.
```
What are "upstream bootstrap configurations"?

onur-ozkan replied:
build configurations / config.toml. Will make it more clear.

RalfJung replied:
Is that the same as "profiles"?

onur-ozkan replied:
No, profiles doesn't include/cover the entire `config.toml` options. But `config.toml` covers profiles too.

onur-ozkan replied:
> Used when upstream bootstrap configurations change

e.g., changing the default value of any option in `config.example.toml`, or directly in the bootstrap configuration deserializer.

RalfJung replied:
So that sounds like "Used when default values change"?

onur-ozkan replied:
"Used when default values change or options removed in build configurations."

What about that? I think "or options removed" is worth to mention.

RalfJung replied:
Yeah sure. Though "removed in build configurations" is again unclear, since you didn't remove them in the user's config. What about

"Used when the default value of an option changes, or support for an option is removed entirely"

onur-ozkan replied:
https://github.com/rust-lang/rust/pull/117815/commits/06f6cd95beca124a54a7346ea2e31b75e1bcfc5a

onur-ozkan committed:
bootstrap: add doc-comments for `ChangeSeverity`

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented:
-kindly reminder ping- @albertlarsan68

albertlarsan68 reviewed (approved):
LGTM, sorry for the delay

albertlarsan68 commented:
Thanks for the PR!
@bors r+

bors commented:
:pushpin: Commit 9a6afd0362acee1ee9685f50e6c5286a19baf6d9 has been approved by `albertlarsan68`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=albertlarsan68 9a6afd0362acee1ee9685f50e6c5286a19baf6d9 -->
<!-- homu: {"type":"Approved","sha":"9a6afd0362acee1ee9685f50e6c5286a19baf6d9","approver":"albertlarsan68","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 9a6afd0362acee1ee9685f50e6c5286a19baf6d9 with merge 91f7f266ce973a08e6e700f9ce815f7b61d347e9...
<!-- homu: {"type":"BuildStarted","head_sha":"9a6afd0362acee1ee9685f50e6c5286a19baf6d9","merge_sha":"91f7f266ce973a08e6e700f9ce815f7b61d347e9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6969965228/job/18971372317)
Approved by: albertlarsan68
Pushing 91f7f266ce973a08e6e700f9ce815f7b61d347e9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"albertlarsan68","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6969965228/job/18971372317"},"merge_sha":"91f7f266ce973a08e6e700f9ce815f7b61d347e9"} -->

rust-timer commented:
Finished benchmarking commit (91f7f266ce973a08e6e700f9ce815f7b61d347e9): [comparison URL](https://perf.rust-lang.org/compare.html?start=e7b22851a2c97980fd4e8235f3255af4fe0a12d5&end=91f7f266ce973a08e6e700f9ce815f7b61d347e9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e7b22851a2c97980fd4e8235f3255af4fe0a12d5&end=91f7f266ce973a08e6e700f9ce815f7b61d347e9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e7b22851a2c97980fd4e8235f3255af4fe0a12d5&end=91f7f266ce973a08e6e700f9ce815f7b61d347e9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.6% | [3.7%, 5.6%] | 2     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.6% | [3.7%, 5.6%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e7b22851a2c97980fd4e8235f3255af4fe0a12d5&end=91f7f266ce973a08e6e700f9ce815f7b61d347e9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=e7b22851a2c97980fd4e8235f3255af4fe0a12d5&end=91f7f266ce973a08e6e700f9ce815f7b61d347e9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 676.441s -> 677.025s (0.09%)
**Artifact size**: 313.75 MiB -> 313.73 MiB (-0.01%)
